os/crypto/blake2s.c

```diff
@@ -0,0 +1,588 @@
+    G(r,3,v[ 3],v[ 7],v[11],v[15]); \
+    G(r,4,v[ 0],v[ 5],v[10],v[15]); \
+    G(r,5,v[ 1],v[ 6],v[11],v[12]); \
+    G(r,6,v[ 2],v[ 7],v[ 8],v[13]); \
+    G(r,7,v[ 3],v[ 4],v[ 9],v[14]); \
+  } while(0)
+
+	/* Size vs performance trade-off. With unrolling, on ARMv7-M function text
+	 * is ~4 KiB and without ~1 KiB. Without unrolling we take ~25% performance
+	 * hit. */
+
+#if 1
+	/* Smaller, slightly slower. */
+
+	for (round = 0; round < 10; round++) {
+		ROUND(round);
+	}
+#else
+	/* Larger, slightly faster. */
+
+	(void)(round = 0);
+	ROUND(0);
+	ROUND(1);
+	ROUND(2);
+	ROUND(3);
+	ROUND(4);
+	ROUND(5);
+	ROUND(6);
+	ROUND(7);
+	ROUND(8);
+	ROUND(9);
+#endif
+
+#undef G
+#undef ROUND
+
+	for (i = 0; i < 8; ++i) {
+		S->h[i] = S->h[i] ^ v[i] ^ v[i + 8];
+	}
+}
+
+#ifdef CONFIG_BLAKE2_SELFTEST
+/* BLAKE2s self-test from RFC 7693 */
+
+static void selftest_seq(FAR uint8_t *out, size_t len, uint32_t seed)
+{
+	size_t i;
+	uint32_t t;
+	uint32_t a;
+	uint32_t b;
+
+	a = 0xDEAD4BAD * seed;		/* prime */
+	b = 1;
+	/* fill the buf */
+	for (i = 0; i < len; i++) {
+		t = a + b;
+		a = b;
+		b = t;
+		out[i] = (t >> 24) & 0xFF;
+	}
+}
+
+static int blake2s_selftest(void)
+{
+	/* Grand hash of hash results. */
+
+	static const uint8_t blake2s_res[32] = {
+		0x6a, 0x41, 0x1f, 0x08, 0xce, 0x25, 0xad, 0xcd, 0xfb, 0x02, 0xab, 0xa6,
+		0x41, 0x45, 0x1c, 0xec, 0x53, 0xc5, 0x98, 0xb2, 0x4f, 0x4f, 0xc7, 0x87,
+		0xfb, 0xdc, 0x88, 0x79, 0x7f, 0x4c, 0x1d, 0xfe
+	};
+
+	/* Parameter sets. */
+
+	static const size_t b2s_md_len[4] = { 16, 20, 28, 32 };
+	static const size_t b2s_in_len[6] = { 0, 3, 64, 65, 255, 1024 };
+	size_t i;
+	size_t j;
+	size_t outlen;
+	size_t inlen;
+	FAR uint8_t *in;
+	uint8_t md[32];
+	uint8_t key[32];
+	blake2s_state ctx;
+	int ret = -1;
+
+	in = malloc(1024);
+	if (!in) {
+		goto out;
+	}
+
+	/* 256-bit hash for testing. */
+
+	if (blake2s_init(&ctx, 32)) {
+		goto out;
+	}
+
+	for (i = 0; i < 4; i++) {
+		outlen = b2s_md_len[i];
+		for (j = 0; j < 6; j++) {
+			inlen = b2s_in_len[j];
+
+			selftest_seq(in, inlen, inlen);	/* unkeyed hash */
+			blake2s(md, outlen, in, inlen, NULL, 0);
+			blake2s_update(&ctx, md, outlen);	/* hash the hash */
+
+			selftest_seq(key, outlen, outlen);	/* keyed hash */
+			blake2s(md, outlen, in, inlen, key, outlen);
+			blake2s_update(&ctx, md, outlen);	/* hash the hash */
+		}
+	}
+
+	/* Compute and compare the hash of hashes. */
+
+	blake2s_final(&ctx, md, 32);
+	for (i = 0; i < 32; i++) {
+		if (md[i] != blake2s_res[i]) {
+			goto out;
+		}
+	}
+
+	ret = 0;
+
+out:
+	free(in);
+	return ret;
+}
+#endif
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/* init2 xors IV with input parameter block */
+
+int blake2s_init_param(FAR blake2s_state *S, FAR const blake2s_param *P)
+{
+	FAR const unsigned char *p = (FAR const unsigned char *)(P);
```

Comment:
Please check for validity of pointers before using them, especially when the function is non static.